### PR TITLE
Remove matplotlibrc.template.

### PR DESCRIPTION
**Pull request by anntzer**
## PR Summary

See https://github.com/matplotlib/matplotlib/pull/19588, https://github.com/matplotlib/matplotlib/pull/19602.

## PR Checklist

- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [ ] New features are documented, with examples if plot related.
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

**Comment by jklymak** [comment]
We can close #19602 if simpler, though suggest stealing some of the commentary.  Also suggest the softlink so no one complains...

**Comment by jklymak** [comment]
### More detail:

Previously `matplotlibrc.template` was changed in #15029 to be the canonical place to get the defaults for all the rcParams.  However, these were *read* during `__init__.py` from `lib/matplotlib/mpl-data/matplotlibrc`.  `lib/matplotlib/mpl-data/matplotlibrc` doesn't exist in the repo, and only gets created when `setup.py` is called.  We can't read the defaults from `matplotlbrc.template` because that file probably doesn't exist in most user installs.  

Aside from being somewhat rickety, the problem is that developers working from an editable install don't want to always call `setup.py` as it requires compiling the C extensions, and hence their `lib/matplotlib/mpl-data/matplotlibrc` can get out of sync with the library.    

The solution here is to simply make `lib/matplotlib/mpl-data/matplotlibrc` the canonical source of the defaults, add it to version control, and softlink it to `matplotlib.template`.  `setup.py` can change `lib/matplotlib/mpl-data/matplotlibrc` in-place (I think)

See also:
- #11844
- https://github.com/matplotlib/matplotlib/pull/14929
- https://github.com/matplotlib/matplotlib/pull/15029

**Review by jklymak — APPROVED** [review]
This seems fine to me, though I do not fully grok the setup.py changes.

File: setup.py [hunk 1]
```diff
@@ -200,9 +203,37 @@ def build_extensions(self):
...
 
-cmdclass = versioneer.get_cmdclass()
+def update_matplotlibrc(path):
+    # Update the matplotlibrc file if packagers want to change the default
+    # backend.
+    template_lines = path.read_text().splitlines(True)
+    backend_line_idx, = [  # Also asserts that there is a single such line.
+        idx for idx, line in enumerate(template_lines)
+        if line.startswith("#backend:")]
+    if setupext.options["backend"]:
```
Comment (timhoffm) [inline setup.py:213]: Can we get into trouble if this had been set once but has been removed again? In that case we would not revert to the original value.

Comment (anntzer) [reply setup.py:213]: indeed

File: setup.py [hunk 2]
```diff
@@ -200,9 +203,38 @@ def build_extensions(self):
...
+    # Update the matplotlibrc file if packagers want to change the default
+    # backend.
+    template_lines = path.read_text().splitlines(True)
+    backend_line_idx, = [  # Also asserts that there is a single such line.
+        idx for idx, line in enumerate(template_lines)
+        if line.startswith("#backend:")]
+    template_lines[backend_line_idx] = (
+        "#backend: {}".format(setupext.options["backend"])
+        if setupext.options["backend"]
+        else "#backend:")
```
Comment (tacaswell) [inline setup.py:216]: The template currently say

```
#backend: Agg
```

so this is not a no-op in any case.  I think we should change the checked-in matplotlibrc file to match this.

It took me a bit to sort out again, but the backend default is explicitly not inferred from this file (as it defaults to a sentinel object).

**Review by tacaswell — APPROVED** [review]
modulo making sure it is a no-op in the default case.

**Comment by anntzer** [comment]
added a test checking the default backend.

**Pull request merged**